src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,161 @@
+            return false;
+        }
+        for (T listItem : list) {
+            add(listItem);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (first == null) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+        }
+        return searchNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> currentNode = searchNodeByIndex(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = searchNodeByIndex(index);
+        T oldValue = currentNode.value;
+        unlink(currentNode);
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = first;
+        while (currentNode != null) {
+            if (currentNode.value == object || object != null && object.equals(currentNode.value)) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+        }
+    }
+
+    private void unlink(Node<T> currentNode) {
+        Node<T> previousNode = currentNode.prev;
+        Node<T> nextNode = currentNode.next;
+        if (previousNode == null) {
+            first = nextNode;
+        } else {
+            previousNode.next = nextNode;
+            currentNode.prev = null;
+        }
+        if (nextNode == null) {
+            last = previousNode;
+        } else {
+            nextNode.prev = previousNode;
+            currentNode.next = null;
+        }
+        size--;
+    }
+
+    private Node<T> searchNodeByIndex(int index) {
+        checkIndex(index);
+        Node<T> currentNode;
+        int i;
+        if (index <= size / 2) {
+            currentNode = first;
+            i = 0;
+            while (i != index) {
+                currentNode = currentNode.next;
+                i++;
+            }
+        } else {
+            currentNode = last;
+            i = size - 1;
+            while (i != index) {
+                currentNode = currentNode.prev;
+                i--;
+            }
+        }
+        return currentNode;
+    }
+
+    private class Node<E> {
+        private Node<E> prev;
+        private Node<E> next;
+        private E value;
+
+        public Node(E value) {
+            this.value = value;
+        }
     }
```

Comment:
can we put this class upper so we can read code easier?
maube we can add all Noda fields in consructor?

Reply:
I don't agree. I think, that class Node should be the last, after all private methods

Reply:
ok

Reply:
I think that Konstantin is right